### PR TITLE
[Resolver] Give a meaningful message for the case where there is no stable version for a pod, and user do not explicitly specify the required version.

### PR DESCRIPTION
**Pull request by manuyavuz**
As stated in #4197, `pod install` and `pod update` commands print wrong error information for the following case:

- There are only pre-release versions available for a pod specified in the Podfile (no stable versions).
- User does not explicitly set a version for the pod. She just uses `pod 'PodName'`

This PR fixes the case by  providing a meaningful error message for the occurring dependency conflict. 

While handling resolver error in `handle_resolver_error` method of `resolver.rb` file, the code will check if the specification that is not satisfying given requirement was a pre-release version, and if the requirement was including any explicit version or source. If spec is a pre-release version, and there is no explicit version or source for the requirement, the code will print a message directing the user to specify pre-release version explicitly like this:

```
[!] There is no corresponding stable version for `PodName`. 
You should explicitly specify the version in order to install a pre-release version of `PodName`
```

However, I'm not fully sure that the proposed check handles all scenarios. A double check from anyone would be appreciated.

Thanks!

### CODE REVIEW

**Comment by segiddins** [comment]
Looks good! Just needs specs and a CHANGELOG entry

**Comment by orta** [comment]
Yeah, nice message @manuyavuz  :+1:

**Comment by manuyavuz** [comment]
Thanks for the feedback! I'll update CHANGELOG now. 

I have a question, travis failed with the following offense:

lib/cocoapods/resolver.rb:372:5: C: Perceived complexity for handle_resolver_error is too high. [15/10]
    def handle_resolver_error(error)
    ^^^

Do you have any suggestions on splitting this method in order to reduce complexity?

**Comment by segiddins** [comment]
Nah, we should be ignoring that in the rubocop config anyways, we prefer to judge complexity manually

**Comment by manuyavuz** [comment]
At last! My fingers were resisting to use 2 space indentations :)

I've also added to specs for `resolver.rb`. Only thing fails is 'perceived complexity' thing.

I think it can be merged now?

**Comment by segiddins** [comment]
I'll take a look when I'm next at a computer, but thanks so much @manuyavuz !

**Comment by manuyavuz** [comment]
Hi again, 

My previous implementation was not complete. It was not giving a good message for the following sample case:

`pod 'AFNetworking', '< 1.0', '> 0.10.1'`

For the above scenario, there are three possibilities (1.0RC3, 1.0RC2, 1.0RC1) all of which are pre-release versions. Previous message would be misleading. Here is the new message for the above case:

```
[!] There are only pre-release versions available satisfying the following requirements:

	'AFNetworking', '< 1.0, > 0.10.1'

You should explicitly specify the version in order to install a pre-release version
```

**Comment by orta** [comment]
Clever :+1:

**Comment by segiddins** [comment]
Brilliant! I'll merge this if you can rebase it! :+1: 
Thanks @manuyavuz!

**Comment by manuyavuz** [comment]
Rebase completed! 👋🏽

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,12 @@ To install release candidates run `[sudo] gem install cocoapods --pre`
 ##### Bug Fixes
 
 * Use watchsimulator when validating pods with the watchOS platform.
```
Comment (segiddins) [inline CHANGELOG.md]: this looks extraneous, perhaps a rebase error?

Comment (manuyavuz) [inline CHANGELOG.md]: Yes, seems like auto-merge have failed. Will fix now.

**Comment by segiddins** [comment]
:ship: thanks so much!

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,7 +8,12 @@ To install release candidates run `[sudo] gem install cocoapods --pre`
 
 ##### Bug Fixes
 
-* Use watchsimulator when validating pods with the watchOS platform.  
+* Give a meaningful message for the case where there is no available stable version for a pod,
+  and user do not explicitly specify the required version.
```
Comment (mrackwitz) [inline CHANGELOG.md]: Add two-spaces at the end of the line for a line break.

File: CHANGELOG.md [hunk 3]
```diff
@@ -8,7 +8,12 @@ To install release candidates run `[sudo] gem install cocoapods --pre`
 
 ##### Bug Fixes
 
-* Use watchsimulator when validating pods with the watchOS platform.
```
Comment (mrackwitz) [inline CHANGELOG.md]: You / presumably your editor accidentally removed the two spaces here.

**Comment by segiddins** [comment]
@mrackwitz I just clean that up when doing the release

**Comment by manuyavuz** [comment]
@mrackwitz I removed these because git diff --check was giving 'trailing spaces' warning. I did not think that two spaces needed for the line break. Is it a need for GitHub Markdown format, or required by default markdown format?

**Comment by mrackwitz** [comment]
@manuyavuz: It's a general markdown thing, not specific to GitHub.
Nice work, btw. :+1:

**Comment by manuyavuz** [comment]
I'll keep that in mind, thanks for the remind! :+1:

**Pull request merged**